File: web/src/engine/utils/genai-math.ts
// =============================================================================
// web/src/engine/utils/genai-math.ts
//
// Mathematical utility functions for Generative AI algorithm generators.
// Every function is PURE — no side effects, no mutable state.
//
// All operations use IEEE 754 double precision (JavaScript `number`).
// Numerical stability is prioritized over raw performance.
//
// CRITICAL: These functions implement the EXACT formulae from Section 4 of
// the Phase 8 specification. Any deviation is a bug.
// =============================================================================

// ─────────────────────────────────────────────────────────────────────────────
// 1. Deterministic PRNG
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Creates a seeded pseudo-random number generator based on a string seed.
 * Uses a simple mulberry32 PRNG for deterministic, reproducible sequences.
 *
 * The hash function converts the seed string to a 32-bit integer, and the
 * PRNG generates floats in [0, 1) from that seed.
 *
 * @param seed — Any string. Same seed always produces the same sequence.
 * @returns A function that returns the next pseudo-random number in [0, 1).
 */
export function seedRandom(seed: string): () => number {
  // Convert string to 32-bit hash using djb2.
  let hash = 5381;
  for (let i = 0; i < seed.length; i++) {
    hash = ((hash << 5) + hash + seed.charCodeAt(i)) | 0;
  }

  // Mulberry32 PRNG.
  let state = hash | 0;
  return () => {
    state = (state + 0x6d2b79f5) | 0;
    let t = Math.imul(state ^ (state >>> 15), 1 | state);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

// ─────────────────────────────────────────────────────────────────────────────
// 2. Matrix Operations
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Multiplies two 2D matrices.
 *
 * C[i][j] = Σ_{k=0}^{p-1} A[i][k] × B[k][j]
 *
 * @param A — Matrix of shape [m, p].
 * @param B — Matrix of shape [p, n].
 * @returns Matrix C of shape [m, n].
 * @throws If inner dimensions do not match.
 */
export function matMul(A: number[][], B: number[][]): number[][] {
  const m = A.length;
  const p = A[0]?.length ?? 0;
  const p2 = B.length;
  const n = B[0]?.length ?? 0;

  if (p !== p2) {
    throw new Error(
      `matMul dimension mismatch: A is [${m}, ${p}], B is [${p2}, ${n}]. ` +
        `Inner dimensions ${p} and ${p2} must match.`,
    );
  }

  const C: number[][] = Array.from({ length: m }, () => new Array<number>(n).fill(0));

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      let sum = 0;
      for (let k = 0; k < p; k++) {
        sum += A[i]![k]! * B[k]![j]!;
      }
      C[i]![j] = sum;
    }
  }

  return C;
}

/**
 * Transposes a 2D matrix.
 *
 * result[j][i] = M[i][j]
 *
 * @param M — Matrix of shape [m, n].
 * @returns Transposed matrix of shape [n, m].
 */
export function transpose(M: number[][]): number[][] {
  const m = M.length;
  const n = M[0]?.length ?? 0;
  const result: number[][] = Array.from({ length: n }, () => new Array<number>(m).fill(0));
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      result[j]![i] = M[i]![j]!;
    }
  }
  return result;
}

/**
 * Scales every element of a matrix by a scalar factor.
 *
 * result[i][j] = M[i][j] × scalar
 *
 * @param M      — Input matrix.
 * @param scalar — Scaling factor.
 * @returns New matrix with same shape.
 */
export function scaleMatrix(M: number[][], scalar: number): number[][] {
  return M.map((row) => row.map((val) => val * scalar));
}

// ─────────────────────────────────────────────────────────────────────────────
// 3. Softmax
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Computes the softmax of a 1D vector.
 *
 * Uses the numerically stable formulation:
 *   m = max(z)
 *   softmax(z)_i = exp(z_i - m) / Σ exp(z_j - m)
 *
 * POSTCONDITION: sum(result) = 1.0 within ±1e-6.
 * POSTCONDITION: every result[i] >= 0.
 *
 * @param z — Input vector of any length ≥ 1.
 * @returns Probability distribution (same length as z, sums to 1).
 * @throws If z is empty.
 */
export function softmax(z: number[]): number[] {
  if (z.length === 0) throw new Error("softmax: input vector must be non-empty.");

  // Numerical stability: subtract the maximum value before exponentiating.
  const maxVal = Math.max(...z);
  const exps = z.map((val) => Math.exp(val - maxVal));
  const sumExps = exps.reduce((a, b) => a + b, 0);

  // Guard against sum = 0 (happens if all values are -Infinity).
  if (sumExps === 0) {
    return new Array(z.length).fill(1 / z.length);
  }

  return exps.map((e) => e / sumExps);
}

/**
 * Applies softmax to each row of a 2D matrix independently.
 *
 * POSTCONDITION: For each row i: sum(result[i]) = 1.0 within ±1e-6.
 *
 * @param M — Matrix of shape [m, n].
 * @returns Matrix of shape [m, n] where each row is a probability distribution.
 */
export function softmaxRows(M: number[][]): number[][] {
  return M.map((row) => softmax(row));
}

// ─────────────────────────────────────────────────────────────────────────────
// 4. Vector Operations
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Computes the cosine similarity between two vectors.
 *
 * cosineSimilarity = (a · b) / (‖a‖ × ‖b‖)
 *
 * @param a — First vector.
 * @param b — Second vector. Must have the same length as a.
 * @returns Cosine similarity in [-1, 1]. Returns 0 if either vector is zero.
 * @throws If vectors have different lengths.
 */
export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length) {
    throw new Error(
      `cosineSimilarity: vectors must have same length. Got ${a.length} and ${b.length}.`,
    );
  }

  let dot = 0;
  let normASq = 0;
  let normBSq = 0;

  for (let i = 0; i < a.length; i++) {
    dot += a[i]! * b[i]!;
    normASq += a[i]! * a[i]!;
    normBSq += b[i]! * b[i]!;
  }

  const normA = Math.sqrt(normASq);
  const normB = Math.sqrt(normBSq);

  // Guard against zero vectors.
  if (normA < 1e-12 || normB < 1e-12) return 0;

  return dot / (normA * normB);
}

/**
 * Adds two vectors element-wise.
 *
 * result[i] = a[i] + b[i]
 *
 * @throws If vectors have different lengths.
 */
export function vectorAdd(a: number[], b: number[]): number[] {
  if (a.length !== b.length) {
    throw new Error(`vectorAdd: length mismatch (${a.length} vs ${b.length}).`);
  }
  return a.map((val, i) => val + b[i]!);
}

// ─────────────────────────────────────────────────────────────────────────────
// 5. Layer Normalization
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Applies layer normalization to a 1D vector (simplified, no gamma/beta).
 *
 * output[i] = (x[i] - mean) / sqrt(variance + epsilon)
 *
 * Where:
 *   mean     = (1/d) × Σ x[i]
 *   variance = (1/d) × Σ (x[i] - mean)²
 *   epsilon  = 1e-5
 *
 * POSTCONDITION: mean(output) ≈ 0 (within ±1e-6).
 * POSTCONDITION: variance(output) ≈ 1 (within ±1e-4).
 *
 * @param x — Input vector.
 * @returns Normalized vector (same length).
 */
export function layerNorm(x: number[]): number[] {
  const d = x.length;
  if (d === 0) return [];

  const EPSILON = 1e-5;

  const mean = x.reduce((sum, val) => sum + val, 0) / d;
  const variance = x.reduce((sum, val) => sum + (val - mean) ** 2, 0) / d;
  const stdDev = Math.sqrt(variance + EPSILON);

  return x.map((val) => (val - mean) / stdDev);
}

/**
 * Applies layer normalization to each row of a matrix.
 */
export function layerNormRows(M: number[][]): number[][] {
  return M.map((row) => layerNorm(row));
}

// ─────────────────────────────────────────────────────────────────────────────
// 6. ReLU Activation
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Applies ReLU (Rectified Linear Unit) element-wise.
 *
 * ReLU(x) = max(0, x)
 */
export function relu(x: number[]): number[] {
  return x.map((val) => Math.max(0, val));
}

/**
 * Applies ReLU element-wise to a 2D matrix.
 */
export function reluMatrix(M: number[][]): number[][] {
  return M.map((row) => relu(row));
}

// ─────────────────────────────────────────────────────────────────────────────
// 7. Feed-Forward Network
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Applies a two-layer feed-forward network to each row of the input.
 *
 * FFN(x) = ReLU(x × W_1 + b_1) × W_2 + b_2
 *
 * @param X   — Input matrix [seqLen, d_model].
 * @param W1  — First weight matrix [d_model, d_ff].
 * @param b1  — First bias vector [d_ff].
 * @param W2  — Second weight matrix [d_ff, d_model].
 * @param b2  — Second bias vector [d_model].
 * @returns Output matrix [seqLen, d_model].
 */
export function feedForward(
  X: number[][],
  W1: number[][],
  b1: number[],
  W2: number[][],
  b2: number[],
): number[][] {
  // Step 1: X × W_1
  let hidden = matMul(X, W1);

  // Step 2: + b_1 (broadcast bias to each row)
  hidden = hidden.map((row) => row.map((val, j) => val + b1[j]!));

  // Step 3: ReLU
  hidden = reluMatrix(hidden);

  // Step 4: × W_2
  let output = matMul(hidden, W2);

  // Step 5: + b_2 (broadcast bias to each row)
  output = output.map((row) => row.map((val, j) => val + b2[j]!));

  return output;
}

// ─────────────────────────────────────────────────────────────────────────────
// 8. Embedding and Weight Generation Helpers
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Generates deterministic embedding vectors from token strings.
 *
 * Each token's embedding is derived from its string content using a seeded PRNG,
 * ensuring reproducible results across runs.
 *
 * @param tokens — Array of token strings.
 * @param dim    — Embedding dimension.
 * @returns Matrix of shape [tokens.length, dim].
 */
export function generateEmbeddings(tokens: string[], dim: number): number[][] {
  return tokens.map((token) => {
    const rng = seedRandom(`embedding-${token}`);
    return Array.from({ length: dim }, () => Math.round((rng() * 2 - 1) * 1000) / 1000);
  });
}

/**
 * Generates a deterministic weight matrix from a seed name.
 *
 * Values are in [-0.5, 0.5] to mimic Xavier initialization scale.
 *
 * @param name — Seed name (e.g., "W_Q", "W_K", "W_V").
 * @param rows — Number of rows.
 * @param cols — Number of columns.
 * @returns Matrix of shape [rows, cols].
 */
export function generateWeightMatrix(name: string, rows: number, cols: number): number[][] {
  const rng = seedRandom(`weight-${name}`);
  return Array.from({ length: rows }, () =>
    Array.from({ length: cols }, () => Math.round((rng() - 0.5) * 1000) / 1000),
  );
}

/**
 * Generates a deterministic bias vector from a seed name.
 *
 * Values are small (near zero) to mimic standard initialization.
 *
 * @param name — Seed name (e.g., "b_1", "b_2").
 * @param length — Vector length.
 * @returns Vector of the given length.
 */
export function generateBiasVector(name: string, length: number): number[] {
  const rng = seedRandom(`bias-${name}`);
  return Array.from({ length }, () => Math.round((rng() - 0.5) * 200) / 1000);
}

File: python/src/eigenvue/math_utils/genai_math.py
"""
GenAI Mathematics Utility Library — Python mirror of genai-math.ts.

Pure functions for all mathematical operations used by generative AI generators.

CRITICAL PARITY CONTRACT:
Every function in this module MUST produce results identical to its
TypeScript counterpart in ``web/src/engine/utils/genai-math.ts``.
The cross-language parity tests enforce this with +/-1e-9 tolerance
for floating-point values.

KEY DESIGN RULES:
1. Every function is PURE (no side effects, no mutable state except the PRNG).
2. Every function is DETERMINISTIC given the same PRNG seed.
3. The PRNG algorithm is Mulberry32 with DJB2 string hash, matching the
   TypeScript implementation EXACTLY.
4. NO use of Python's ``random`` module. ALL randomness comes from seed_random().
5. NO use of NumPy. This module uses plain Python arithmetic for exact
   parity with TypeScript's plain-JavaScript arithmetic.

MATHEMATICAL OPERATIONS:
- Matrix multiplication (mat_mul)
- Matrix transpose (transpose)
- Row-wise softmax with numerical stability (softmax_rows)
- Scalar matrix scaling (scale_matrix)
- Cosine similarity (cosine_similarity)
- L2 norm (l2_norm)
- Dot product (dot_product)
- Vector addition (vector_add)
- Layer normalization (layer_norm, layer_norm_rows)
- ReLU activation (relu_vec, relu_matrix)
- Feed-forward network (feed_forward)
- Deterministic embedding generation (generate_embeddings)
- Deterministic weight matrix generation (generate_weight_matrix)
- Deterministic bias vector generation (generate_bias_vector)
"""

from __future__ import annotations

import math
from collections.abc import Callable

# ── Seeded PRNG ──────────────────────────────────────────────────────────────
# PARITY CRITICAL: Must match the TypeScript implementation in genai-math.ts.
# TypeScript uses:
#   1. DJB2 hash to convert string seed -> 32-bit integer
#   2. Mulberry32 PRNG for generating floats in [0, 1)
# Python's integers are arbitrary precision, so we must use bitwise masks
# to simulate 32-bit signed integer overflow behavior.


def _to_int32(x: int) -> int:
    """Convert a Python integer to a signed 32-bit integer (JS `| 0` behavior).

    Parameters
    ----------
    x : int
        Any Python integer.

    Returns
    -------
    int
        Value in [-2^31, 2^31 - 1].
    """
    x = x & 0xFFFFFFFF
    if x >= 0x80000000:
        return x - 0x100000000
    return x


def _to_uint32(x: int) -> int:
    """Convert a Python integer to an unsigned 32-bit integer (JS `>>> 0`).

    Parameters
    ----------
    x : int
        Any Python integer.

    Returns
    -------
    int
        Value in [0, 2^32 - 1].
    """
    return x & 0xFFFFFFFF


def _math_imul(a: int, b: int) -> int:
    """Emulate JavaScript's Math.imul() — 32-bit integer multiplication.

    Parameters
    ----------
    a : int
        First operand.
    b : int
        Second operand.

    Returns
    -------
    int
        Signed 32-bit result of a * b.
    """
    # Perform multiplication and truncate to 32 bits
    return _to_int32(a * b)


def seed_random(seed: str) -> Callable[[], float]:
    """Create a seeded pseudo-random number generator.

    PARITY CRITICAL: This MUST match the TypeScript ``seedRandom()`` in
    ``genai-math.ts`` exactly. The TypeScript implementation uses:
    1. DJB2 hash to convert the string seed to a 32-bit integer.
    2. Mulberry32 PRNG to generate floats in [0, 1).

    Parameters
    ----------
    seed : str
        Any string seed. Same seed always produces the same sequence.

    Returns
    -------
    Callable[[], float]
        A function that returns the next pseudo-random float in [0, 1)
        each time it is called.

    Examples
    --------
    >>> rng = seed_random("test")
    >>> val = rng()
    >>> 0.0 <= val < 1.0
    True
    """
    # Step 1: DJB2 hash — convert string to 32-bit integer
    # PARITY CRITICAL: Must match TypeScript's djb2 hash exactly
    hash_val = 5381
    for i in range(len(seed)):
        # hash = ((hash << 5) + hash + seed.charCodeAt(i)) | 0
        hash_val = _to_int32((hash_val << 5) + hash_val + ord(seed[i]))

    # Step 2: Mulberry32 PRNG
    # PARITY CRITICAL: Must match TypeScript's mulberry32 exactly
    state = [_to_int32(hash_val)]

    def next_float() -> float:
        """Return the next pseudo-random float in [0, 1)."""
        # state = (state + 0x6d2b79f5) | 0
        state[0] = _to_int32(state[0] + 0x6D2B79F5)
        # let t = Math.imul(state ^ (state >>> 15), 1 | state)
        t = _math_imul(state[0] ^ (_to_uint32(state[0]) >> 15), 1 | state[0])
        # t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t
        t = _to_int32(t + _math_imul(t ^ (_to_uint32(t) >> 7), 61 | t)) ^ t
        # return ((t ^ (t >>> 14)) >>> 0) / 4294967296
        return _to_uint32(t ^ (_to_uint32(t) >> 14)) / 4294967296

    return next_float


# ── Matrix Operations ────────────────────────────────────────────────────────


def mat_mul(a: list[list[float]], b: list[list[float]]) -> list[list[float]]:
    """Matrix multiplication: C = A x B.

    Uses naive O(m*k*n) triple-loop for exact parity with TypeScript.

    Parameters
    ----------
    a : list[list[float]]
        Matrix A with shape [m, k].
    b : list[list[float]]
        Matrix B with shape [k, n].

    Returns
    -------
    list[list[float]]
        Result matrix C with shape [m, n].
        C[i][j] = Sigma_{p=0}^{k-1} A[i][p] * B[p][j]

    Raises
    ------
    ValueError
        If inner dimensions do not match (A's columns != B's rows).
    """
    m = len(a)
    if m == 0:
        return []

    k = len(a[0])
    if k == 0:
        return [[] for _ in range(m)]

    if len(b) != k:
        raise ValueError(
            f"mat_mul: inner dimensions mismatch — A has {k} columns but B has {len(b)} rows."
        )

    n = len(b[0]) if k > 0 else 0

    # C[i][j] = Sigma_p A[i][p] * B[p][j]
    result: list[list[float]] = []
    for i in range(m):
        row: list[float] = []
        for j in range(n):
            total = 0.0
            for p in range(k):
                total += a[i][p] * b[p][j]
            row.append(total)
        result.append(row)

    return result


def transpose(matrix: list[list[float]]) -> list[list[float]]:
    """Transpose a 2D matrix.

    Parameters
    ----------
    matrix : list[list[float]]
        Input matrix with shape [m, n].

    Returns
    -------
    list[list[float]]
        Transposed matrix with shape [n, m].
        result[j][i] = matrix[i][j]
    """
    if not matrix:
        return []
    m = len(matrix)
    n = len(matrix[0])
    result: list[list[float]] = [[0.0] * m for _ in range(n)]
    for i in range(m):
        for j in range(n):
            result[j][i] = matrix[i][j]
    return result


def scale_matrix(matrix: list[list[float]], scalar: float) -> list[list[float]]:
    """Multiply every element of a matrix by a scalar.

    Parameters
    ----------
    matrix : list[list[float]]
        Input matrix.
    scalar : float
        Scaling factor.

    Returns
    -------
    list[list[float]]
        New matrix where result[i][j] = matrix[i][j] * scalar.
    """
    return [[val * scalar for val in row] for row in matrix]


# ── Softmax ──────────────────────────────────────────────────────────────────


def softmax(z: list[float]) -> list[float]:
    """Compute the softmax of a 1D vector.

    Uses the numerically stable formulation:
        softmax(x_i) = exp(x_i - max(x)) / Sigma_j exp(x_j - max(x))

    INVARIANT: sum(result) = 1.0 within +/-1e-9. All values >= 0.

    Parameters
    ----------
    z : list[float]
        Input vector of any length >= 1.

    Returns
    -------
    list[float]
        Probability distribution (same length as z, sums to 1).

    Raises
    ------
    ValueError
        If z is empty.
    """
    if len(z) == 0:
        raise ValueError("softmax: input vector must be non-empty.")

    # Numerical stability: subtract the maximum value before exponentiating
    max_val = max(z)
    exps = [math.exp(val - max_val) for val in z]
    sum_exps = sum(exps)

    # Guard against sum = 0 (happens if all values are -Infinity)
    if sum_exps == 0:
        n = len(z)
        return [1.0 / n] * n

    return [e / sum_exps for e in exps]


def softmax_rows(matrix: list[list[float]]) -> list[list[float]]:
    """Apply softmax independently to each row of a matrix.

    INVARIANT: For each row i: abs(sum(row) - 1.0) <= 1e-9.

    Parameters
    ----------
    matrix : list[list[float]]
        Input matrix with shape [m, n].

    Returns
    -------
    list[list[float]]
        Matrix of same shape where each row is a probability distribution.
    """
    return [softmax(row) for row in matrix]


# ── Vector Operations ────────────────────────────────────────────────────────


def dot_product(a: list[float], b: list[float]) -> float:
    """Compute the dot product of two vectors.

    Parameters
    ----------
    a : list[float]
        First vector.
    b : list[float]
        Second vector (must be same length as ``a``).

    Returns
    -------
    float
        Scalar dot product: Sigma_i a[i] * b[i]

    Raises
    ------
    ValueError
        If vectors have different lengths.
    """
    if len(a) != len(b):
        raise ValueError(f"dot_product: vectors must be same length ({len(a)} vs {len(b)}).")
    total = 0.0
    for i in range(len(a)):
        total += a[i] * b[i]
    return total


def l2_norm(vec: list[float]) -> float:
    """Compute the L2 (Euclidean) norm of a vector.

    Parameters
    ----------
    vec : list[float]
        Input vector.

    Returns
    -------
    float
        ||vec||_2 = sqrt(Sigma_i vec[i]^2)
    """
    return math.sqrt(sum(v * v for v in vec))


def cosine_similarity(a: list[float], b: list[float]) -> float:
    """Compute cosine similarity between two vectors.

    cos_sim(a, b) = (a . b) / (||a||_2 x ||b||_2)

    Parameters
    ----------
    a : list[float]
        First vector.
    b : list[float]
        Second vector (same length as ``a``).

    Returns
    -------
    float
        Cosine similarity in [-1, 1]. Returns 0.0 if either vector
        has near-zero norm.
    """
    if len(a) != len(b):
        raise ValueError(
            f"cosine_similarity: vectors must have same length. Got {len(a)} and {len(b)}."
        )
    dot = 0.0
    norm_a_sq = 0.0
    norm_b_sq = 0.0
    for i in range(len(a)):
        dot += a[i] * b[i]
        norm_a_sq += a[i] * a[i]
        norm_b_sq += b[i] * b[i]

    norm_a = math.sqrt(norm_a_sq)
    norm_b = math.sqrt(norm_b_sq)

    # Guard against zero vectors (matches TypeScript threshold)
    if norm_a < 1e-12 or norm_b < 1e-12:
        return 0.0

    return dot / (norm_a * norm_b)


def vector_add(a: list[float], b: list[float]) -> list[float]:
    """Add two vectors element-wise.

    Parameters
    ----------
    a : list[float]
        First vector.
    b : list[float]
        Second vector (must be same length).

    Returns
    -------
    list[float]
        result[i] = a[i] + b[i]

    Raises
    ------
    ValueError
        If vectors have different lengths.
    """
    if len(a) != len(b):
        raise ValueError(f"vector_add: length mismatch ({len(a)} vs {len(b)}).")
    return [a[i] + b[i] for i in range(len(a))]


# ── Layer Normalization ──────────────────────────────────────────────────────


def layer_norm(x: list[float]) -> list[float]:
    """Apply layer normalization to a 1D vector (simplified, no gamma/beta).

    output[i] = (x[i] - mean) / sqrt(variance + epsilon)

    Where:
        mean     = (1/d) x Sigma x[i]
        variance = (1/d) x Sigma (x[i] - mean)^2
        epsilon  = 1e-5

    Parameters
    ----------
    x : list[float]
        Input vector.

    Returns
    -------
    list[float]
        Normalized vector (same length).
    """
    d = len(x)
    if d == 0:
        return []

    epsilon = 1e-5
    mean = sum(x) / d
    variance = sum((val - mean) ** 2 for val in x) / d
    std_dev = math.sqrt(variance + epsilon)

    return [(val - mean) / std_dev for val in x]


def layer_norm_rows(matrix: list[list[float]]) -> list[list[float]]:
    """Apply layer normalization to each row of a matrix.

    Parameters
    ----------
    matrix : list[list[float]]
        Input matrix.

    Returns
    -------
    list[list[float]]
        Normalized matrix.
    """
    return [layer_norm(row) for row in matrix]


# ── ReLU Activation ──────────────────────────────────────────────────────────


def relu_vec(x: list[float]) -> list[float]:
    """Apply ReLU element-wise to a vector.

    ReLU(x) = max(0, x)

    Parameters
    ----------
    x : list[float]
        Input vector.

    Returns
    -------
    list[float]
        Vector with ReLU applied.
    """
    return [max(0.0, val) for val in x]


def relu_matrix(matrix: list[list[float]]) -> list[list[float]]:
    """Apply ReLU element-wise to a 2D matrix.

    Parameters
    ----------
    matrix : list[list[float]]
        Input matrix.

    Returns
    -------
    list[list[float]]
        Matrix with ReLU applied.
    """
    return [relu_vec(row) for row in matrix]


# ── Feed-Forward Network ─────────────────────────────────────────────────────


def feed_forward(
    x: list[list[float]],
    w1: list[list[float]],
    b1: list[float],
    w2: list[list[float]],
    b2: list[float],
) -> list[list[float]]:
    """Apply a two-layer feed-forward network.

    FFN(x) = ReLU(x x W_1 + b_1) x W_2 + b_2

    Parameters
    ----------
    x : list[list[float]]
        Input matrix [seq_len, d_model].
    w1 : list[list[float]]
        First weight matrix [d_model, d_ff].
    b1 : list[float]
        First bias vector [d_ff].
    w2 : list[list[float]]
        Second weight matrix [d_ff, d_model].
    b2 : list[float]
        Second bias vector [d_model].

    Returns
    -------
    list[list[float]]
        Output matrix [seq_len, d_model].
    """
    # Step 1: X x W_1
    hidden = mat_mul(x, w1)

    # Step 2: + b_1 (broadcast bias to each row)
    hidden = [[val + b1[j] for j, val in enumerate(row)] for row in hidden]

    # Step 3: ReLU
    hidden = relu_matrix(hidden)

    # Step 4: x W_2
    output = mat_mul(hidden, w2)

    # Step 5: + b_2 (broadcast bias to each row)
    output = [[val + b2[j] for j, val in enumerate(row)] for row in output]

    return output


# ── Embedding and Weight Generation Helpers ──────────────────────────────────


def generate_embeddings(tokens: list[str], dim: int) -> list[list[float]]:
    """Generate deterministic pseudo-random embeddings for a list of tokens.

    PARITY CRITICAL: Must match TypeScript ``generateEmbeddings()`` exactly.
    Each token's embedding is generated using seedRandom(`embedding-{token}`)
    and values are rounded to 3 decimal places: Math.round((rng() * 2 - 1) * 1000) / 1000.

    Parameters
    ----------
    tokens : list[str]
        List of token strings.
    dim : int
        Embedding dimension.

    Returns
    -------
    list[list[float]]
        Matrix of shape [len(tokens), dim]. Each row is a token's embedding.
    """
    result: list[list[float]] = []
    for token in tokens:
        rng = seed_random(f"embedding-{token}")
        embedding = [round((rng() * 2 - 1) * 1000) / 1000 for _ in range(dim)]
        result.append(embedding)
    return result


def generate_weight_matrix(name: str, rows: int, cols: int) -> list[list[float]]:
    """Generate a deterministic pseudo-random weight matrix.

    PARITY CRITICAL: Must match TypeScript ``generateWeightMatrix()`` exactly.
    Uses seedRandom(`weight-{name}`) and values are rounded:
    Math.round((rng() - 0.5) * 1000) / 1000.

    Parameters
    ----------
    name : str
        Matrix identifier (e.g., "W_Q", "W_K", "W_V").
    rows : int
        Number of rows.
    cols : int
        Number of columns.

    Returns
    -------
    list[list[float]]
        Matrix of shape [rows, cols] with values in [-0.5, 0.5].
    """
    rng = seed_random(f"weight-{name}")
    matrix: list[list[float]] = []
    for _ in range(rows):
        row = [round((rng() - 0.5) * 1000) / 1000 for _ in range(cols)]
        matrix.append(row)
    return matrix


def generate_bias_vector(name: str, length: int) -> list[float]:
    """Generate a deterministic pseudo-random bias vector.

    PARITY CRITICAL: Must match TypeScript ``generateBiasVector()`` exactly.
    Uses seedRandom(`bias-{name}`) and values are:
    Math.round((rng() - 0.5) * 200) / 1000.

    Parameters
    ----------
    name : str
        Bias identifier (e.g., "b_1", "b_2").
    length : int
        Vector length.

    Returns
    -------
    list[float]
        Bias vector.
    """
    rng = seed_random(f"bias-{name}")
    return [round((rng() - 0.5) * 200) / 1000 for _ in range(length)]
